### PR TITLE
Fix #84 Apply url encode to the request url

### LINKED ISSUE #84: Fail to get object with key contains space
The master version seems to have a bug with when save object with key contains space.

```
Downloading data/Permanent Files/LoyaltyHist.csv to data_tmp/data/Permanent Files/LoyaltyHist.csv                                   
$Code                                                                                                                               
[1] "HttpVersionNotSupported"                                                                                                       
                                                                                                                                    
$Message                                                                                                                            
[1] "The HTTP version specified is not supported."
```

I did google around and also look at the source, it seems the cause is from [this commit](https://github.com/cloudyr/aws.s3/commit/d9f62daf4f04ee1d4e497a9c582df2798b905bdc). I think we need to escape the URL as well, not only the action.

I'm double check and will submit a PR if this is valid.

### PR DESCRIPTION
**Pull request by nqbao**
This will fix #84 by applyling url encode to the request url.

### CODE REVIEW

**Comment by nqbao** [comment]
@leeper please take a look and let me know if i need to change anything

**Comment by nqbao** [comment]
This seems to have problem with put_object. Let me check further

**Comment by nqbao** [comment]
Okie, this should be good now.

File: R/s3HTTP.R [hunk 1]
```diff
@@ -55,19 +55,19 @@ s3HTTP <- function(verb = "GET",
...
+        ), collapse = '/')
     }
 
     url <- setup_s3_url(bucketname, region, encodedPath, accelerate)
     p <- parse_url(url)
 
     current <- Sys.time()
     d_timestamp <- format(current, "%Y%m%dT%H%M%SZ", tz = "UTC")
-    action <- if (p$path == "") "/" else p$path
+    action <- if (p$path == "") "/" else paste0("/", p$path)
```
Comment (nqbao) [inline R/s3HTTP.R]: I made this change because httr::parse_url will not return path with `/` prefix

File: R/s3HTTP.R [hunk 2]
```diff
@@ -53,21 +53,24 @@ s3HTTP <- function(verb = "GET",
...
     }
-    url <- setup_s3_url(bucketname, region, path, accelerate)
-    p <- parse_url(url)
     
-    current <- Sys.time()
-    d_timestamp <- format(current, "%Y%m%dT%H%M%SZ", tz = "UTC")
-    action <- if (p$path == "") "/" else {
-        paste0("/", paste(sapply(
-            strsplit(p$path, '/')[[1]],
+    encodedPath <- if (path == "") "/" else {
```
Comment (nqbao) [inline R/s3HTTP.R]: path should always contain `/` prefix

**Comment by rnorberg** [comment]
I was just going through the codebase to resolve this error and I traced it back to `URLendode()`ing the `url` passed to `GET()`/`DELETE()`/`POST()`/`PUT()`. I confirmed that your fix resolves issue #84 . Perhaps add a test for this?

```
test_that("putobject and deleteobject handle object names with spaces and special characters", {
  tmp <- tempfile(pattern = 'tricky file name &$@=:+,? ', fileext = ".txt")
  writeLines(c("cloudyr", "test"), tmp)
  
  p <- put_object(
    file = tmp,
    object = basename(tmp),
    bucket = 'hpk', 
    key = Sys.getenv("TRAVIS_AWS_ACCESS_KEY_ID"), 
    secret = Sys.getenv("TRAVIS_AWS_SECRET_ACCESS_KEY")
  )
  expect_true(p)
  
  d <- delete_object(
    object = basename(tmp),
    bucket = 'hpk',
    key = Sys.getenv("TRAVIS_AWS_ACCESS_KEY_ID"), 
    secret = Sys.getenv("TRAVIS_AWS_SECRET_ACCESS_KEY")
  )
  expect_true(d)
  
  if (file.exists(tmp)) file.remove(tmp)
})
```

**Comment by nqbao** [comment]
I will add that test.

**Comment by nqbao** [comment]
Okie it's green now. Look like we are good to go.

**Comment by nqbao** [comment]
is it good to merge now @rnorberg

**Pull request merged**